core/server/data/migrations/versions/5.0/2022-03-28-15-25-backfill-mrr-adjustments-for-offers.js

```diff
@@ -0,0 +1,109 @@
+            .from('offer_redemptions AS or')
+            .join('offers AS o', 'or.offer_id', '=', 'o.id')
+            .join('members_stripe_customers_subscriptions AS s', 'or.subscription_id', '=', 's.id')
+            .join('stripe_prices AS p', 'p.stripe_price_id', '=', 's.stripe_price_id')
+            .where('o.duration', '=', 'forever')
+            .whereNotNull('s.offer_id')
+            .where('s.mrr', '!=', 0)
+            .orderBy('or.created_at', 'asc');
+
+        if (offerRedemptions.length === 0) {
+            logging.info('No Offers redeemed, skipping migration');
+            return;
+        } else {
+            logging.info(`Adjusting MRR for ${offerRedemptions.length} Offer Redemptions`);
+        }
+
+        const memberIds = uniq(offerRedemptions.map(redemption => redemption.member_id));
+
+        const mrrCreatedEvents = await knex
+            .select('*')
+            .from('members_paid_subscription_events')
+            .where('type', 'created')
+            .whereIn('member_id', memberIds);
+
+        function storeEventOnMemberId(storage, event) {
+            return {
+                ...storage,
+                [event.member_id]: storage[event.member_id] ? storage[event.member_id].concat(event) : [event]
+            };
+        }
+
+        const mrrCreatedEventsByMemberId = mrrCreatedEvents.reduce(storeEventOnMemberId, {});
+
+        const updatedEvents = [];
+
+        function calculateMRR(subscription, redemption) {
+            if (redemption && subscription.interval !== redemption.discount_interval) {
+                logging.error('Found invalid price & redemption pair');
+                logging.info(redemption.subscription_id);
+                return calculateMRR(subscription);
+            }
+
+            if (!redemption) {
+                return subscription.interval === 'year' ? subscription.amount / 12 : subscription.amount;
+            }
+
+            if (redemption.discount_type === 'percent') {
+                return calculateMRR({
+                    interval: subscription.interval,
+                    amount: subscription.amount * (100 - redemption.discount_amount) / 100
+                });
+            }
+
+            return calculateMRR({
+                interval: subscription.interval,
+                amount: subscription.amount - redemption.discount_amount
+            });
+        }
+
+        offerRedemptions.forEach((redemption) => {
+            const memberEvents = mrrCreatedEventsByMemberId[redemption.member_id];
+
+            // If a member has had multiple subscriptions we ignore because we cannot easily work out which event is correct.
+            if (memberEvents.length !== 1) {
+                return;
+            }
+
+            const firstEvent = memberEvents[0];
+
+            const mrr = calculateMRR({
+                interval: redemption.interval,
+                amount: redemption.amount
+            }, redemption);
+            logging.info(mrr);
```

Comment:
I suspect this is left over from debugging

Reply:
Yup! Will fix